boot/setefibootvars_linux.go

```diff
@@ -0,0 +1,197 @@
+	ErrInvalidBootOrder   = errors.New("BootOrder variable data must have even length")
+
+	defaultVarAttrs = efi.AttributeNonVolatile | efi.AttributeBootserviceAccess | efi.AttributeRuntimeAccess
+
+	efiListVariables = efi.ListVariables
+	efiReadVariable  = efi.ReadVariable
+	efiWriteVariable = efi.WriteVariable
+
+	linuxFilePathToDevicePath = linux.FilePathToDevicePath
+
+	bootOptionRegexp = regexp.MustCompile("^Boot[0-9A-F]{4}$")
+)
+
+// constructLoadOption returns a serialized EFI load option with the device
+// path corresponding to the given asset path, along with the given description
+// and optional data.
+func constructLoadOption(description string, assetPath string, optionalData []byte) ([]byte, error) {
+	devicePath, err := linuxFilePathToDevicePath(assetPath, linux.ShortFormPathHD)
+	if err != nil {
+		return nil, err
+	}
+	loadOption := &efi.LoadOption{
+		Attributes:   efi.LoadOptionActive | efi.LoadOptionCategoryBoot,
+		Description:  description,
+		FilePath:     devicePath,
+		OptionalData: optionalData,
+	}
+	loadOptionSerialized, err := loadOption.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	return loadOptionSerialized, nil
+}
+
+// Searches existing Boot#### variables for one which matches the given data.
+//
+// If there is a match, returns the boot number of the existing variable.
+// Otherwise, finds the first available boot number and returns it, along with
+// ErrNoMatchingVariable, indicating that a new boot option with that boot
+// number should be written. If a different error occurs, returns that error,
+// and the returned boot number should be ignored.
+func findMatchingBootOption(optionData []byte) (uint16, error) {
+	variables, err := efiListVariables()
+	if err != nil {
+		return 0, err
+	}
+	usedBootNums := make(map[uint64]bool)
+	for _, varDesc := range variables {
+		varName := varDesc.Name
+		varGUID := varDesc.GUID
+		if !bootOptionRegexp.MatchString(varName) {
+			// Not a Boot#### variable, so skip it
+			continue
+		}
+		if varGUID != efi.GlobalVariable {
+			// Not an EFI global variable, so skip it
+			continue
+		}
+		varNumber, err := strconv.ParseUint(varName[4:], 16, 16)
+		if err != nil {
+			// Should not occur, since variable matched bootOptionRegexp
+			return 0, err
+		}
+		// Since we never overwrite an existing variable, we can ignore
+		// variable attributes when reading the variable
+		varData, _, err := efiReadVariable(varName, varGUID)
+		if err != nil {
+			return 0, err
+		}
+		if bytes.Compare(optionData, varData) == 0 {
+			// existing variable already identical, use it
+			return uint16(varNumber), nil
+		}
+		usedBootNums[varNumber] = true
+	}
+	for bootNum := uint64(0); bootNum <= 0xFFFF; bootNum++ {
+		if !usedBootNums[bootNum] {
+			return uint16(bootNum), ErrNoMatchingVariable
+		}
+	}
+	return 0, ErrAllBootNumsUsed
+}
+
+// Ensures that a Boot#### variable contains the given EFI load option.
```

Comment:
even if not exported it would be good when possible to start with the function name here

Reply:
this comment still applies